### PR TITLE
Revamp exceptions thrown by the dps device - part2

### PR DESCRIPTION
**Pull request by brycewang-microsoft**
+ rename `ProvisioningTransportException` to `DeviceProvisioningClientException` to align naming format of exception thrown by the hub device & service clients
+ add "StatusCode" property into `DeviceProvisioningClientException`
+ remove unused ctors from `DeviceProvisioningClientException`

### CODE REVIEW

File: iothub/device/src/Transport/Http/HttpClientHelper.cs
Comment (brycewang-microsoft) [inline iothub/device/src/Transport/Http/HttpClientHelper.cs:212]: catching and just throwing again is weird, so just not catching anymore

File: provisioning/device/src/DeviceProvisioningClientException.cs
Comment (brycewang-microsoft) [inline provisioning/device/src/DeviceProvisioningClientException.cs:13]: Do we like the naming here `DeviceProvisioningClientException`? Maybe I should change this to `ProvisioningDeviceClientException` to align the naming style for dps device & service clients?

File: iothub/device/src/Exceptions/IotHubClientException.cs
Comment (drwill-ms) [inline iothub/device/src/Exceptions/IotHubClientException.cs:190]: ErrorCode to match service SDK?

File: provisioning/device/src/DeviceProvisioningClientException.cs
Comment (drwill-ms) [reply provisioning/device/src/DeviceProvisioningClientException.cs:13]: Can modules register with DPS, @timtay-microsoft / @abhipsaMisra ?

File: provisioning/device/src/DeviceProvisioningClientException.cs
Comment (drwill-ms) [inline provisioning/device/src/DeviceProvisioningClientException.cs:83]: Odd that on provisioning we have this other class, whereas on hub we have the properties directly on the exception. We should look into making them look the same.

File: iothub/device/src/Transport/Http/HttpClientHelper.cs
Comment (drwill-ms) [reply iothub/device/src/Transport/Http/HttpClientHelper.cs:212]: As long as it doesn't get caught by the Exception catch block below.

File: iothub/device/src/Exceptions/IotHubClientException.cs
Comment (brycewang-microsoft) [reply iothub/device/src/Exceptions/IotHubClientException.cs:190]: No, this is property in `IotHubClientException` thrown by the **hub device**. The changes here is to fix the doc comments only.

Comment (drwill-ms) [reply iothub/device/src/Exceptions/IotHubClientException.cs:190]: And the code will be simply 200, 400, 500? Or the 6-digit kind?

Comment (andyk-ms) [reply iothub/device/src/Exceptions/IotHubClientException.cs:190]: Discussed with Bryce offline and we need to gather the code from DPS which are 6-digits.

Comment (brycewang-microsoft) [reply iothub/device/src/Exceptions/IotHubClientException.cs:190]: > And the code will be simply 200, 400, 500? Or the 6-digit kind?

it's more like 6-digit kind. FYI, I am revamping this `IotHubClientException` to be more aligned with the `IotHubServiceException` by using 2 properties `StatusCode`(3-digit) and `ErrorCode`(6-digit).

Comment (brycewang-microsoft) [reply iothub/device/src/Exceptions/IotHubClientException.cs:190]: "revert" the changes related to "hub client exceptions" in this PR as I am having another PR for `IotHubClientException` in particular.

File: e2e/test/provisioning/ProvisioningE2ETests.cs
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningE2ETests.cs:522]:
```suggestion
            var error = await act.Should().ThrowAsync<DeviceProvisioningClientException>().ConfigureAwait(false);
```

File: provisioning/device/src/DeviceProvisioningClientException.cs
Comment (drwill-ms) [inline provisioning/device/src/DeviceProvisioningClientException.cs:102]: When is this getting called?

**Review by drwill-ms — APPROVED** [review]

File: e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs
Comment (patilsnr) [inline e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs:58]: I'm new to our fluent assertions, but is there a reason we're using `var` instead of an explicit type here? Would the actual type just be too verbose and hard to read?

File: e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs
Comment (patilsnr) [inline e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs:68]: same comment here

**Review by patilsnr — APPROVED** [review]
I like the new name! I'd wait until also getting a sign off from someone who knows this part of our code a bit better than I do, though.

File: e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs
Comment (brycewang-microsoft) [reply e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs:58]: Yes. In other words, it's not necessary to specify the exception type again as we already include it with `ThrowAsync<>`.

File: e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs
Comment (brycewang-microsoft) [reply e2e/test/provisioning/ProvisioningCertificateValidationE2ETest.cs:68]: same as above

**Pull request merged**
